File: dart/lib/src/domain/get_domain_address.dart
import '../constants/addresses.dart';
import '../constants/records.dart';
import '../errors/sns_errors.dart';
import '../utils/derive_address.dart';

/// Result containing derived domain address and metadata.
class DomainAddressResult {
  /// Creates a domain address result.
  const DomainAddressResult({
    required this.domainAddress,
    required this.isSub,
    this.parentAddress,
    this.isSubRecord = false,
  });

  /// The derived on-chain address for the domain.
  final String domainAddress;

  /// The parent domain address (for subdomains).
  final String? parentAddress;

  /// Whether this domain is a subdomain.
  final bool isSub;

  /// Whether this represents a subdomain record.
  final bool isSubRecord;
}

/// Parameters for domain address derivation.
class GetDomainAddressParams {
  /// Creates parameters for domain address derivation.
  const GetDomainAddressParams({
    required this.domain,
    this.record,
  });

  /// The domain name to derive address for (with or without .sol suffix).
  final String domain;

  /// Record version for record-specific addresses (optional).
  final RecordVersion? record;
}

/// Derives the on-chain address for a domain, subdomain, or record.
///
/// Computes the Program Derived Address (PDA) for any .sol domain or subdomain.
/// Supports both standard domains and record-specific addresses.
///
/// Example:
/// ```dart
/// // Get main domain address
/// final result = await getDomainAddress(
///   GetDomainAddressParams(domain: 'bonfida'));
/// print(result.domainAddress); // On-chain address
///
/// // Get subdomain address
/// final subResult = await getDomainAddress(
///   GetDomainAddressParams(domain: 'sub.bonfida'));
/// print(subResult.isSub); // true
/// ```
///
/// [params] Domain name and optional record version.
///
/// Returns [DomainAddressResult] with address and metadata.
Future<DomainAddressResult> getDomainAddress(
    GetDomainAddressParams params) async {
  var domain = params.domain;
  final record = params.record;

  // Remove .sol suffix if present
  if (domain.endsWith('.sol')) {
    domain = domain.substring(0, domain.length - 4);
  }

  final recordClass =
      record == RecordVersion.v2 ? centralStateDomainRecords : null;
  final recordPrefix = _getRecordPrefix(record);
  final splitted = domain.split('.');

  if (splitted.length == 2) {
    // Subdomain case (e.g., "sub.example")
    final parentAddress = await deriveAddress(
      splitted[1],
      parentAddress: rootDomainAddress,
    );
    final domainAddress = await deriveAddress(
      recordPrefix + splitted[0],
      parentAddress: parentAddress,
      classAddress: recordClass,
    );

    return DomainAddressResult(
      domainAddress: domainAddress,
      parentAddress: parentAddress,
      isSub: true,
    );
  } else if (splitted.length == 3 && record != null) {
    // Sub-record case (e.g., "record.sub.example")
    // Parent domain
    final parentAddress = await deriveAddress(
      splitted[2],
      parentAddress: rootDomainAddress,
    );

    // Sub domain
    final subAddress = await deriveAddress(
      '\x00${splitted[1]}',
      parentAddress: parentAddress,
    );

    // Sub record
    final domainAddress = await deriveAddress(
      recordPrefix + splitted[0],
      parentAddress: subAddress,
      classAddress: recordClass,
    );

    return DomainAddressResult(
      domainAddress: domainAddress,
      parentAddress: parentAddress,
      isSub: true,
      isSubRecord: true,
    );
  } else if (splitted.length >= 3) {
    throw InvalidInputError('The domain is malformed');
  }

  // Simple domain case (e.g., "example")
  final domainAddress = await deriveAddress(
    domain,
    parentAddress: rootDomainAddress,
  );

  return DomainAddressResult(
    domainAddress: domainAddress,
    isSub: false,
  );
}

/// Gets the record prefix for a given record version
String _getRecordPrefix(RecordVersion? record) {
  switch (record) {
    case RecordVersion.v2:
      return '\x02';
    case RecordVersion.v1:
      return '\x01';
    case null:
      return '\x00';
  }
}

File: js-kit/src/domain/getDomainAddress.ts
import {
  CENTRAL_STATE_DOMAIN_RECORDS,
  SNS_ROOT_DOMAIN_ACCOUNT,
} from "../constants/addresses";
import { InvalidInputError } from "../errors";
import { RecordVersion } from "../types/record";
import { _deriveAddress } from "../utils/deriveAddress";
import { parseSupportedTld } from "../utils/tld";

interface GetDomainAddressParams {
  domain: string;
  record?: RecordVersion;
}

/**
 * Derives the address of a domain, a subdomain, or a record.
 *
 * @param params - An object containing the following properties:
 *   - `domain`: The full domain name to process, including a .sns or .sol suffix.
 *   - `record`: (Optional) The record version. Only provide if the domain being resolved is a record.
 * @returns A promise that resolves to an object containing the derived address and additional metadata.
 */
export const getDomainAddress = async ({
  domain,
  record,
}: GetDomainAddressParams) => {
  [domain] = parseSupportedTld(domain);

  const recordClass =
    record === RecordVersion.V2 ? CENTRAL_STATE_DOMAIN_RECORDS : undefined;
  const recordPrefix =
    {
      [RecordVersion.V2]: "\x02",
      [RecordVersion.V1]: "\x01",
    }[record as RecordVersion] || "\x00";
  const splitted = domain.split(".");

  if (splitted.length === 2) {
    const parentAddress = await _deriveAddress(
      splitted[1],
      SNS_ROOT_DOMAIN_ACCOUNT
    );
    const domainAddress = await _deriveAddress(
      recordPrefix + splitted[0],
      parentAddress,
      recordClass
    );

    return { domainAddress, parentAddress, isSub: true };
  } else if (splitted.length === 3 && !!record) {
    // Parent domain
    const parentAddress = await _deriveAddress(
      splitted[2],
      SNS_ROOT_DOMAIN_ACCOUNT
    );

    // Sub domain
    const subAddress = await _deriveAddress("\0" + splitted[1], parentAddress);

    // Sub record
    const domainAddress = await _deriveAddress(
      recordPrefix + splitted[0],
      subAddress,
      recordClass
    );

    return { domainAddress, parentAddress, isSub: true, isSubRecord: true };
  } else if (splitted.length >= 3) {
    throw new InvalidInputError("The domain is malformed");
  }

  const domainAddress = await _deriveAddress(domain, SNS_ROOT_DOMAIN_ACCOUNT);

  return { domainAddress, isSub: false };
};
